src/haz3lmenhir/Parser.mly

```diff
@@ -0,0 +1,269 @@
+%token <float> FLOAT
+%token LET
+%token FUN
+%token CASE
+%token OPEN_SQUARE_BRACKET
+%token CLOSE_SQUARE_BRACKET
+%token OPEN_PAREN
+%token CLOSE_PAREN
+%token DASH_ARROW
+%token EQUAL_ARROW
+%token SINGLE_EQUAL
+%token TURNSTILE
+
+(* Int ops *)
+%token DOUBLE_EQUAL
+%token NOT_EQUAL
+%token PLUS
+%token MINUS
+%token DIVIDE
+%token POWER
+%token TIMES
+%token LESS_THAN
+%token LESS_THAN_EQUAL
+%token GREATER_THAN
+%token GREATER_THAN_EQUAL
+(* Float ops *)
+%token DOUBLE_EQUAL_FLOAT
+%token NOT_EQUAL_FLOAT
+%token PLUS_FLOAT
+%token MINUS_FLOAT
+%token DIVIDE_FLOAT
+%token POWER_FLOAT
+%token TIMES_FLOAT
+%token LESS_THAN_FLOAT
+%token LESS_THAN_EQUAL_FLOAT
+%token GREATER_THAN_FLOAT
+%token GREATER_THAN_EQUAL_FLOAT
+(*logical ops*)
+%token L_AND
+%token L_OR
+%token L_NOT
+(*bitwise ops*)
+%token COMMA
+%token COLON
+%token EOF
+%token IN
+%token UNIT
+%token END
+
+(* type tokens *)
+%token INT_TYPE
+%token FLOAT_TYPE
+%token BOOL_TYPE
+%token STRING_TYPE
+%token UNKNOWN
+%token INTERNAL
+
+%token IF
+%token THEN
+%token ELSE
+
+%token SEMI_COLON
+
+
+
+(* Precedences *)
+
+%nonassoc IF_EXP
+%nonassoc LET_EXP
+
+%left DASH_ARROW
+
+%left PLUS MINUS TIMES POWER DIVIDE DOUBLE_EQUAL NOT_EQUAL LESS_THAN_EQUAL GREATER_THAN_EQUAL 
+
+%left GREATER_THAN LESS_THAN
+
+%left PLUS_FLOAT MINUS_FLOAT TIMES_FLOAT POWER_FLOAT DIVIDE_FLOAT DOUBLE_EQUAL_FLOAT NOT_EQUAL_FLOAT LESS_THAN_FLOAT LESS_THAN_EQUAL_FLOAT GREATER_THAN_FLOAT GREATER_THAN_EQUAL_FLOAT
+
+(* Other *)
+%left CONS
+%left OPEN_PAREN
+%left QUESTION
+%left COLON
+%left COMMA
+%left AT_SYMBOL
+%left SEMI_COLON
+%left IN
+%left DOLLAR_SIGN
+%left L_NOT L_AND L_OR
+
+(* Might not be correct - milan *)
+%left IDENT
+
+%type <AST.exp> exp
+
+%start <AST.exp> program
+
+%%
+
+program:
+    | e = exp; EOF {e}
+
+%inline intOp:
+    | PLUS { IntOp(Plus) }
+    | MINUS { IntOp(Minus) }
+    | TIMES { IntOp(Times) }
+    | POWER { IntOp(Power) }
+    | DIVIDE { IntOp(Divide) }
+    | DOUBLE_EQUAL { IntOp(Equals) }
+    | NOT_EQUAL { IntOp(NotEquals) }
+    (* | LESS_THAN { IntOp(LessThan) } *)
+    | LESS_THAN_EQUAL { IntOp(LessThanOrEqual) }
+    | GREATER_THAN { IntOp(GreaterThan) }
+    | GREATER_THAN_EQUAL { IntOp(GreaterThanOrEqual) }
+
+
+%inline floatOp:
+    | PLUS_FLOAT { FloatOp(Plus) }
+    | MINUS_FLOAT { FloatOp(Minus) }
+    | TIMES_FLOAT { FloatOp(Times) }
+    | POWER_FLOAT { FloatOp(Power) }
+    | DIVIDE_FLOAT { FloatOp(Divide) }
+    | DOUBLE_EQUAL_FLOAT { FloatOp(Equals) }
+    | NOT_EQUAL_FLOAT { FloatOp(NotEquals) }
+    | LESS_THAN_FLOAT { FloatOp(LessThan) }
+    | LESS_THAN_EQUAL_FLOAT { FloatOp(LessThanOrEqual) }
+    | GREATER_THAN_FLOAT { FloatOp(GreaterThan) }
+    | GREATER_THAN_EQUAL_FLOAT { FloatOp(GreaterThanOrEqual) }
+
+%inline boolOp:
+    | L_AND { BoolOp(And) }
+    | L_OR { BoolOp(Or) }
+
+%inline binOp:
+    | i = intOp { i }
+    | f = floatOp { f }
+    | b = boolOp { b }
+
+binExp:
+    | e1 = exp; b = binOp; e2 = exp { BinExp (e1, b, e2) }
+
+typ:
```

Comment:
This doesn't currently allow for Typ Variables.

Reply:
@green726 I added https://github.com/hazelgrove/hazel/pull/1220/commits/1aec2b9075548359d9421b143245554fb6fd1090 with some tests for type variable parsing